### PR TITLE
Remove test-only methods from OrderRepository - fixes #3

### LINKED ISSUE #3: Introduce TestOrderResource interface
At the moment we have `clear` method within various `OrderRepository` implementations, which is only used for testing and requires use of reflection, Weld proxy unwrapping, and other ugly hacks.

We should introduce `TestOrderRepository` interface (under `src/main/test`) that extends `OrderRepository` by adding `clear` method (and any other methods we may need to expose for testing) and use it instead in the tests.

### PR DESCRIPTION
**Pull request by olotenko**
Problem: clear() method is used for testing only. Should not exist
       for production use.

Solution: Move it to test classes. This requires reorganizing how
       inheritance works, and how CDI picks up the right implementation
       in tests. @Specializes cannot be used - test implementations
       are seen as ambiguous. Using @Alternative with @Priority appears
       to be the standard way of achieving this. This requires beans.xml
       for all tests.

Issues: fixes #3

### CODE REVIEW

File: orders-core/src/main/java/io/helidon/examples/sockshop/orders/DefaultOrderRepository.java [hunk 1]
```diff
@@ -18,9 +21,11 @@
  * API testing and quick demos.
  */
 @ApplicationScoped
+@Alternative
+@Priority(Interceptor.Priority.APPLICATION+1)
```
Comment (aseovic) [inline orders-core/src/main/java/io/helidon/examples/sockshop/orders/DefaultOrderRepository.java]: Make this `APPLICATION - 10` and remove the `@Alternative` as it is not necessary for the default implementation. We just need to make it lower priority than the other impls.

File: orders-core/src/test/java/io/helidon/examples/sockshop/orders/TestOrderRepository.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package io.helidon.examples.sockshop.orders;
+
+// Tests need to interact with the entity backing OrderRepository
+// to reset to initial state before every test
+public interface TestOrderRepository {
```
Comment (aseovic) [inline orders-core/src/test/java/io/helidon/examples/sockshop/orders/TestOrderRepository.java:22]: Add proper javadoc to both the class and the `clear` method.

File: orders-core/src/test/java/io/helidon/examples/sockshop/orders/TestDefaultOrderRepository.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package io.helidon.examples.sockshop.orders;
+
+import javax.annotation.Priority;
+import javax.enterprise.inject.Alternative;
+import javax.interceptor.Interceptor;
+
+@Alternative
+@Priority(Interceptor.Priority.APPLICATION+10)
```
Comment (aseovic) [inline orders-core/src/test/java/io/helidon/examples/sockshop/orders/TestDefaultOrderRepository.java]: Change priority to `APPLICATION - 5`, so it's higher than the default implementation, but lower than data store-specific implementations (Mongo, Redis, Jpa)

File: orders-mongo/src/main/java/io/helidon/examples/sockshop/orders/mongo/MongoOrderRepository.java [hunk 4]
```diff
@@ -23,11 +25,11 @@
  * that that uses MongoDB as a backend data store.
  */
 @ApplicationScoped
-@Specializes
+@Alternative
+@Priority(Interceptor.Priority.APPLICATION+10)
```
Comment (aseovic) [inline orders-mongo/src/main/java/io/helidon/examples/sockshop/orders/mongo/MongoOrderRepository.java]: Make priority `APPLICATION` -- higher than default and default test implementation.

File: orders-mongo/src/test/java/io/helidon/examples/sockshop/orders/mongo/TestMongoOrderRepository.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import javax.interceptor.Interceptor;
+
+import io.helidon.examples.sockshop.orders.TestOrderRepository;
+import io.helidon.examples.sockshop.orders.Order;
+
+import com.mongodb.client.MongoCollection;
+import org.bson.BsonDocument;
+
+@Alternative
+@Priority(Interceptor.Priority.APPLICATION+100)
```
Comment (aseovic) [inline orders-mongo/src/test/java/io/helidon/examples/sockshop/orders/mongo/TestMongoOrderRepository.java]: `APPLICATION + 5`, to be consistent ;-)

File: orders-mysql/src/main/java/io/helidon/examples/sockshop/orders/jpa/JpaOrderRepository.java [hunk 6]
```diff
@@ -19,12 +22,13 @@
  * that that uses relational database (via JPA) as a backend data store.
  */
 @ApplicationScoped
-@Specializes
+@Alternative
+@Priority(Interceptor.Priority.APPLICATION+10)
```
Comment (aseovic) [inline orders-mysql/src/main/java/io/helidon/examples/sockshop/orders/jpa/JpaOrderRepository.java]: `APPLICATION`

File: orders-mysql/src/test/java/io/helidon/examples/sockshop/orders/jpa/TestJpaOrderRepository.java [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+import javax.annotation.Priority;
+import javax.enterprise.inject.Alternative;
+import javax.interceptor.Interceptor;
+import javax.transaction.Transactional;
+
+import io.helidon.examples.sockshop.orders.TestOrderRepository;
+
+@Alternative
+@Priority(Interceptor.Priority.APPLICATION+100)
```
Comment (aseovic) [inline orders-mysql/src/test/java/io/helidon/examples/sockshop/orders/jpa/TestJpaOrderRepository.java]: `APPLICATION + 5`

File: orders-redis/src/main/java/io/helidon/examples/sockshop/orders/redis/RedisOrderRepository.java [hunk 8]
```diff
@@ -15,11 +20,44 @@
  * that that uses Redis (via Redisson) as a backend data store.
  */
 @ApplicationScoped
-@Specializes
+@Alternative
+@Priority(Interceptor.Priority.APPLICATION+10)
```
Comment (aseovic) [inline orders-redis/src/main/java/io/helidon/examples/sockshop/orders/redis/RedisOrderRepository.java]: `APPLICATION`

File: orders-redis/src/main/java/io/helidon/examples/sockshop/orders/redis/RedisOrderRepository.java [hunk 9]
```diff
@@ -15,11 +20,44 @@
...
 @Traced
-public class RedisOrderRepository extends DefaultOrderRepository {
```
Comment (aseovic) [inline orders-redis/src/main/java/io/helidon/examples/sockshop/orders/redis/RedisOrderRepository.java:20]: The sync one could actually still extend `DefaultOrderRepository`, as it simply changes the underlying storage map to use.

I would split async implementation into `OrderRepositoryAsync/RedisOrderRepositoryAsync`, per example in `Carts` service, and move that change to a separate PR so we don't conflate things.

File: orders-redis/src/test/java/io/helidon/examples/sockshop/orders/redis/TestRedisOrderRepository.java [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+@Alternative
+@Priority(Interceptor.Priority.APPLICATION+100)
+public class TestRedisOrderRepository extends RedisOrderRepository implements TestOrderRepository {
+    @Inject
+    public TestRedisOrderRepository(RMap<String, Order> carts) {
+        super(carts);
+    }
+
+    @Override
+    public void clear() {
```
Comment (aseovic) [inline orders-redis/src/test/java/io/helidon/examples/sockshop/orders/redis/TestRedisOrderRepository.java:22]: This can simply use sync API

**Review by aseovic — CHANGES_REQUESTED** [review]
Overall it looks pretty good, pretty much along the lines of what we discussed.

Let's just change the priorities as suggested so they are consistent across services, and remove async Redis functionality from this PR (that should be a separate PR, and implemented the same way it is implemented for the `Carts` service.

Thanks for your help.

File: orders-redis/src/main/java/io/helidon/examples/sockshop/orders/redis/RedisOrderRepository.java (hunk 9 above)
Comment (olotenko) [reply orders-redis/src/main/java/io/helidon/examples/sockshop/orders/redis/RedisOrderRepository.java:20]: As discussed, I'll do this in this ticket, but we need to revisit how `RedisOrderRepository` is implemented.

File: orders-redis/src/test/java/io/helidon/examples/sockshop/orders/redis/TestRedisOrderRepository.java [hunk 11]
```diff
@@ -11,7 +11,7 @@
 import org.redisson.api.RMap;
 
 @Alternative
-@Priority(Interceptor.Priority.APPLICATION+100)
+@Priority(Interceptor.Priority.APPLICATION+5)
 public class TestRedisOrderRepository extends RedisOrderRepository implements TestOrderRepository {
     @Inject
     public TestRedisOrderRepository(RMap<String, Order> carts) {
```
Comment (aseovic) [inline orders-redis/src/test/java/io/helidon/examples/sockshop/orders/redis/TestRedisOrderRepository.java:17]: Shouldn't the above parameter name be `orders` instead of `carts`?

Comment (olotenko) [reply orders-redis/src/test/java/io/helidon/examples/sockshop/orders/redis/TestRedisOrderRepository.java:17]: Maybe. The origin of the name is in the original `RedisOrderRepository` code in master:

https://github.com/helidon-sock-shop/orders/blob/master/orders-redis/src/main/java/io/helidon/examples/sockshop/orders/redis/RedisOrderRepository.java#L22-L23

Comment (aseovic) [reply orders-redis/src/test/java/io/helidon/examples/sockshop/orders/redis/TestRedisOrderRepository.java:17]: Yeah, definitely copy&paste error on my end. Do you mind fixing it in both places so I can merge it as part of this PR?

Otherwise I can merge as-is and fix it later...
